File: unique-binary-search-trees-ii-95/Solution.java

/**
 * LeetCode problem # 95, 'Unique Binary Search Trees II'
 * https://leetcode.com/problems/unique-binary-search-trees-ii/
 */

import java.util.*;

/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode() {} TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) { this.val = val; this.left
 * = left; this.right = right; } }
 */
class Solution {
    public List<TreeNode> generateTrees(int n) {
        return solve(1, n);
    }

    private List<TreeNode> solve(int start, int end) {
        List<TreeNode> res = new ArrayList<>();
        if (start > end) {
            res.add(null);
            return res;
        }

        for (int i = start; i <= end; ++i) {
            List<TreeNode> left = solve(start, i - 1);
            List<TreeNode> right = solve(i + 1, end);

            for (TreeNode lnode : left) {
                for (TreeNode rnode : right) {
                    TreeNode curr = new TreeNode(i, lnode, rnode);
                    res.add(curr);
                }
            }
        }
        return res;
    }
}

File: unique-binary-search-trees-ii-95/solution.js
/**
 * LeetCode problem # 95, 'Unique Binary Search Trees II'
 * https://leetcode.com/problems/unique-binary-search-trees-ii/
 */

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number} n
 * @return {TreeNode[]}
 */
function generateTrees(n) {
  function solve(start, end) {
    if (start > end) {
      return [null];
    }
    const res = [];
    for (let i = start; i <= end; ++i) {
      const leftTrees = solve(start, i - 1);
      const rightTrees = solve(i + 1, end);

      for (const left of leftTrees) {
        for (const right of rightTrees) {
          const curr = new TreeNode(i);
          curr.left = left;
          curr.right = right;
          res.push(curr);
        }
      }
    }
    return res;
  }
  return n === 0 ? [] : solve(1, n);
}
